### PR TITLE
[GCS] refactor the GCS Client Job Interface

### PR DESCRIPTION
**Pull request by micafan**
## Why are these changes needed?

Refactoring the GCS Client Job Interface. Only expose Ray-specific semantics (e.g., `AsyncSubscribe`, `AsyncUpdate`, etc), and hide the implementation details about the backed storage. 

For more details please ref the proposal [Refactor the GCS Client](https://docs.google.com/document/d/1WHH5CjHFQYro0KSqSRyBQx282mfHhZ1GBri-mpS3IlE/edit#).

## Related issue number

Closes #5058

## Linter

- [X] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by jovany-wang** [comment]
@micafan Is this ready to review?

**Comment by micafan** [comment]
> @micafan Is this ready to review?

yes.

File: src/ray/common/pb_util.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <memory>
+#include "ray/common/id.h"
+#include "ray/protobuf/gcs.pb.h"
+
+namespace ray {
+
+/// Helper function to produce job table data (for newly created job or updated job).
+///
+/// \param job_id The ID of job that need to register or update.
```
Comment (jovany-wang) [inline src/ray/common/pb_util.h]:
```suggestion
/// \param job_id The ID of job that needs to be registered or updated.
```

File: src/ray/common/pb_util.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/// \param is_dead Whether the job is dead.
```
Comment (jovany-wang) [inline src/ray/common/pb_util.h]:
```suggestion
/// \param is_dead Whether the driver of this job is dead.
```

File: src/ray/common/pb_util.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+/// \param timestamp The UNIX timestamp of corresponding to this event.
+/// \param node_manager_address Address of node this job was started on.
```
Comment (jovany-wang) [inline src/ray/common/pb_util.h]:
```suggestion
/// \param node_manager_address Address of the node this job was started on.
```

File: src/ray/gcs/subscription_executor.h [hunk 4]
```diff
@@ -71,10 +71,10 @@ class SubscriptionExecutor {
   bool registered_{false};
 
   /// Subscribe Callback of all elements.
-  SubscribeCallback<ID, Data> subscribe_all_callback_{nullptr};
+  SubscribePairCallback<ID, Data> subscribe_all_callback_{nullptr};
 
   /// A mapping from element ID to subscription callback.
-  typedef std::unordered_map<ID, SubscribeCallback<ID, Data>> IDToCallbackMap;
+  typedef std::unordered_map<ID, SubscribePairCallback<ID, Data>> IDToCallbackMap;
```
Comment (jovany-wang) [inline src/ray/gcs/subscription_executor.h]:
```suggestion
using IDToCallbackMap = std::unordered_map<ID, SubscribePairCallback<ID, Data>>;
```

File: src/ray/gcs/redis_gcs_client.h [hunk 5]
```diff
@@ -19,7 +19,11 @@ class RedisContext;
 
 class RAY_EXPORT RedisGcsClient : public GcsClientInterface {
   friend class ActorStateAccessor;
+  friend class JobStateAccessor;
   friend class SubscriptionExecutorTest;
+  friend void TestLogSubscribeAll(const JobID &, std::shared_ptr<gcs::RedisGcsClient>);
+  friend void TestLogSubscribeId(const JobID &, std::shared_ptr<gcs::RedisGcsClient>);
+  friend void TestLogSubscribeCancel(const JobID &, std::shared_ptr<gcs::RedisGcsClient>);
```
Comment (jovany-wang) [inline src/ray/gcs/redis_gcs_client.h]: It's not a good practice to declare a `Test` method as a `friend` in my sense, but I'm not sure.

@zhijunfu @pcmoritz  How do you think of this?

File: src/ray/gcs/actor_state_accessor.cc [hunk 6]
```diff
@@ -91,17 +91,37 @@ Status ActorStateAccessor::AsyncUpdate(const ActorID &actor_id,
...
+  auto on_subscribe = [subscribe](const ActorID &actor_id,
+                                  const std::vector<ActorTableData> &data) {
+    if (!data.empty()) {
+      // We only need the last entry, because it represents the latest state of
+      // this actor.
+      subscribe(actor_id, data.back());
+    }
+  };
+
+  if (actor_id.IsNil()) {
```
Comment (jovany-wang) [inline src/ray/gcs/actor_state_accessor.cc]: Leave a comment to explain about this `if`.

File: src/ray/gcs/actor_state_accessor.cc [hunk 7]
```diff
@@ -91,17 +91,37 @@ Status ActorStateAccessor::AsyncUpdate(const ActorID &actor_id,
 }
 
 Status ActorStateAccessor::AsyncSubscribe(
```
Comment (jovany-wang) [inline src/ray/gcs/actor_state_accessor.cc]: How about rename this method to `AsyncSubscribeAll`?

File: src/ray/gcs/job_state_accessor.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  /// \param data_ptr The job that will be registered or updated to GCS.
+  /// \param callback Callback that will be called after job has been registered
+  /// or updated to GCS.
+  /// \return Status
+  Status DoAsyncRegisterOrUpdate(const std::shared_ptr<JobTableData> &data_ptr,
+                                 const StatusCallback &callback);
+
+  RedisGcsClient &client_impl_;
+
+  typedef SubscriptionExecutor<JobID, std::vector<JobTableData>, JobTable>
```
Comment (jovany-wang) [inline src/ray/gcs/job_state_accessor.h]: `using` ?

File: src/ray/gcs/redis_gcs_client.cc [hunk 9]
```diff
@@ -112,34 +112,38 @@ Status RedisGcsClient::Connect(boost::asio::io_service &io_service) {
...
-  actor_checkpoint_table_.reset(new ActorCheckpointTable(shard_contexts_, this));
-  actor_checkpoint_id_table_.reset(new ActorCheckpointIdTable(shard_contexts_, this));
-  resource_table_.reset(new DynamicResourceTable({primary_context_}, this));
-
-  actor_accessor_.reset(new ActorStateAccessor(*this));
-
   Status status = Attach(io_service);
   is_connected_ = status.ok();
 
+  if (is_connected_) {
```
Comment (jovany-wang) [inline src/ray/gcs/redis_gcs_client.cc]: This line seems always be true right?

**Review by jovany-wang — COMMENTED** [review]
Thanks, I left several minor comments.

File: src/ray/gcs/redis_gcs_client.h (hunk 5 above)
Comment (zhijunfu) [reply src/ray/gcs/redis_gcs_client.h]: This looks a bit strange to me too... Would it work if we instead put the test class as friend?

Comment (micafan) [reply src/ray/gcs/redis_gcs_client.h]: Actually those test didn't belong to any test class.

File: src/ray/gcs/actor_state_accessor.cc (hunk 6 above)
Comment (micafan) [reply src/ray/gcs/actor_state_accessor.cc]: done.

File: src/ray/gcs/actor_state_accessor.cc (hunk 7 above)
Comment (micafan) [reply src/ray/gcs/actor_state_accessor.cc]: How about rename this method in another PR, then we can keep this PR clean.

File: src/ray/gcs/redis_gcs_client.cc (hunk 9 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.cc]: status.ok() return  bool.

File: src/ray/common/pb_util.h (hunk 1 above)
Comment (micafan) [reply src/ray/common/pb_util.h]: done

File: src/ray/gcs/redis_gcs_client.h (hunk 5 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.h]: Maybe I can move those tests to a new test Class.

Comment (micafan) [reply src/ray/gcs/redis_gcs_client.h]: Already add test helper class `LogSubscribeTestHelper` as friend class of `RedisGcsClient`.
Please let me known if you guys has any suggestion.

File: src/ray/common/pb_util.h [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#ifndef RAY_COMMON_PB_UTIL_H
+#define RAY_COMMON_PB_UTIL_H
```
Comment (zhijunfu) [inline src/ray/common/pb_util.h]: The `CreateJobTableData` is only called by `NodeManager`, should we put it to the `raylet` folder instead? 

If it's indeed a util function and will be called by other places, could we instead rename this file to `src/ray/util/gcs_util.h`? The name `pb_util.h` seems ambiguous.

File: src/ray/raylet/node_manager.cc [hunk 11]
```diff
@@ -1025,11 +1023,12 @@ void NodeManager::ProcessDisconnectClientMessage(
   } else if (is_driver) {
     // The client is a driver.
     const auto job_id = worker->GetAssignedJobId();
-    const auto driver_id = ComputeDriverIdFromJob(job_id);
     RAY_CHECK(!job_id.IsNil());
-    RAY_CHECK_OK(gcs_client_->job_table().AppendJobData(
-        job_id, /*is_dead=*/true, std::time(nullptr),
-        initial_config_.node_manager_address, worker->Pid()));
+    auto job_info_ptr =
```
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: job_data_ptr?

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -849,9 +846,10 @@ void NodeManager::ProcessRegisterClientRequestMessage(
     status = worker_pool_.RegisterDriver(std::move(worker));
     if (status.ok()) {
       local_queues_.AddDriverTaskId(driver_task_id);
-      RAY_CHECK_OK(gcs_client_->job_table().AppendJobData(
-          job_id, /*is_dead=*/false, std::time(nullptr),
-          initial_config_.node_manager_address, message->worker_pid()));
+      auto job_info_ptr =
```
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: job_data_ptr since it's obtained from CreateJobTableData?

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -224,13 +224,11 @@ ray::Status NodeManager::RegisterGcs() {
       /*subscribe_callback=*/nullptr,
       /*done_callback=*/nullptr));
 
-  // Subscribe to driver table updates.
-  const auto job_table_handler = [this](gcs::RedisGcsClient *client, const JobID &job_id,
-                                        const std::vector<JobTableData> &job_data) {
-    HandleJobTableUpdate(job_id, job_data);
+  // Subscribe to job updates.
+  const auto job_sub_handler = [this](const std::vector<JobTableData> &job_data) {
```
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]:
```suggestion
  const auto job_subscribe_handler = [this](const std::vector<JobTableData> &job_data) {
```

File: src/ray/gcs/subscription_executor.cc [hunk 14]
```diff
@@ -132,7 +127,8 @@ Status SubscriptionExecutor<ID, Data, Table>::AsyncUnsubscribe(
   return table_.CancelNotifications(JobID::Nil(), id, client_id, on_done);
 }
 
-template class SubscriptionExecutor<ActorID, ActorTableData, ActorTable>;
+template class SubscriptionExecutor<ActorID, std::vector<ActorTableData>, ActorTable>;
+template class SubscriptionExecutor<JobID, std::vector<JobTableData>, JobTable>;
```
Comment (zhijunfu) [inline src/ray/gcs/subscription_executor.cc]: simple question: why do we need the history for job table data (and thus requires a vector instead of a single entry)?

**Review by zhijunfu — COMMENTED** [review]
Thanks, I left a few comments.

File: src/ray/gcs/subscription_executor.cc (hunk 14 above)
Comment (micafan) [reply src/ray/gcs/subscription_executor.cc]: I miss understand the method `NodeManager::HandleJobTableUpdate`, this method will process every record of this job vector.  So if is ok for this method just got the last record of this vector?

File: src/ray/common/pb_util.h (hunk 10 above)
Comment (micafan) [reply src/ray/common/pb_util.h]: The `CreateJobTableData` is also used in GCS test.
 It's better to put this file to the `common` folder rather than GCS, because there may be more helper methods for protobuf in future.

Comment (raulchen) [reply src/ray/common/pb_util.h]: This `CreateJobTableData` function is specific to "job table of GCS", so I agree that it makes more sense to put it under `src/ray/gcs/util.h` and built into the `gcs` `cc_library`.
 
Regarding the common util functions for protobuf, we already have a `grpc_util.h`, which handles common data structures of protobuf. By "common", I mean not specific to any concrete `.proto` definitions. For those concrete proto definitions, better just put in their own modules.

File: src/ray/gcs/job_state_accessor.h [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+  ~JobStateAccessor() {}
+
+  /// Register a job to GCS asynchronously.
+  ///
+  /// \param data_ptr The job that will be registered to GCS.
+  /// \param callback Callback that will be called after job has been registered
+  /// to GCS.
+  /// \return Status
+  Status AsyncRegister(const std::shared_ptr<JobTableData> &data_ptr,
```
Comment (raulchen) [inline src/ray/gcs/job_state_accessor.h]: what about naming this `AsyncAddJob`, whose meaning is clearer than `register`?
Also, we can just pass in `job_id`, `timestamp`, `node_manager_address` and `driver_pid`, because `is_dead` is already `false` and we want to hide the implementation details about `table`.

File: src/ray/gcs/job_state_accessor.h [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+                       const StatusCallback &callback);
+
+  /// Update dynamic states of job in GCS asynchronously.
+  ///
+  /// \param data_ptr The job that will be updated to GCS.
+  /// \param callback Callback that will be called after update finishes.
+  /// \return Status
+  Status AsyncUpdate(const std::shared_ptr<JobTableData> &data_ptr,
```
Comment (raulchen) [inline src/ray/gcs/job_state_accessor.h]: This can be `AsyncMarkJobAsFinished(job_id, callback)`, because this is the only valid update operation.

File: src/ray/gcs/job_state_accessor.h [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+                     const StatusCallback &callback);
+
+  /// Subscribe to any update operations of jobs.
+  ///
+  /// \param subscribe Callback that will be called each time when an job is registered
+  /// or updated.
+  /// \param done Callback that will be called when subscription is complete.
+  /// \return Status
+  Status AsyncSubscribeAll(const SubscribeCallback<JobID, JobTableData> &subscribe,
```
Comment (raulchen) [inline src/ray/gcs/job_state_accessor.h]: Same reason as above, this can be `AsyncSubscribeToDeadJobs`.

File: src/ray/gcs/redis_gcs_client.cc (hunk 9 above)
Comment (raulchen) [inline src/ray/gcs/redis_gcs_client.cc]: Is this change related to this PR? Or does this just fix an unrelated existing issue?

File: src/ray/gcs/redis_gcs_client.h [hunk 18]
```diff
@@ -75,8 +76,10 @@ class RAY_EXPORT RedisGcsClient : public GcsClientInterface {
   /// one event loop should be attached at a time.
   Status Attach(boost::asio::io_service &io_service);
 
-  /// Use method Actors() instead
+  /// Use method Actors() instead.
   ActorTable &actor_table();
+  /// Use method Jobs() instead.
```
Comment (raulchen) [inline src/ray/gcs/redis_gcs_client.h]: what does this comment mean? Is it a TODO?

File: BUILD.bazel [hunk 19]
```diff
@@ -685,6 +685,17 @@ cc_binary(
...
+cc_binary(
+    name = "job_state_accessor_test",
+    testonly = 1,
+    srcs = ["src/ray/gcs/job_state_accessor_test.cc"],
+    copts = COPTS,
+    deps = [
+        ":gcs",
+        "@com_google_googletest//:gtest_main",
+    ],
+)
```
Comment (raulchen) [inline BUILD.bazel]: As discussed before, we should change this to a `cc_test`.  But we can do that in the next PR.

File: src/ray/gcs/job_state_accessor_test.cc [hunk 20]
```diff
@@ -0,0 +1,85 @@
+#include "ray/gcs/job_state_accessor.h"
```
Comment (raulchen) [inline src/ray/gcs/job_state_accessor_test.cc]: the new convention is to put this file in `src/ray/gcs/test/job_state_accessor_test.cc`

File: src/ray/common/pb_util.h (hunk 10 above)
Comment (micafan) [reply src/ray/common/pb_util.h]: We already have `util/util.h`, it's kind of weird if we put it under src/ray/gcs/util.h`. Better to rename it  to `src/ray/gcs/pb_util.h`.

File: src/ray/gcs/job_state_accessor.h (hunk 15 above)
Comment (micafan) [reply src/ray/gcs/job_state_accessor.h]: > what about naming this `AsyncAddJob`, whose meaning is clearer than `register`?

I think `add` it's ok.  But `add` it's more suitable for adding data, what if gcs is a service not only a storage in future? There might be some other logic not only adding data.

> Also, we can just pass in `job_id`, `timestamp`, `node_manager_address` and `driver_pid`, because `is_dead` is already `false` and we want to hide the implementation details about `table`.

I think it's better use a struct rather than list all parameters.  Because if we add a parameter in future, we won't need change the interface.

File: src/ray/gcs/job_state_accessor.h (hunk 16 above)
Comment (micafan) [reply src/ray/gcs/job_state_accessor.h]: Done

File: src/ray/gcs/job_state_accessor.h (hunk 17 above)
Comment (micafan) [reply src/ray/gcs/job_state_accessor.h]: done

File: src/ray/gcs/job_state_accessor_test.cc (hunk 20 above)
Comment (micafan) [reply src/ray/gcs/job_state_accessor_test.cc]: got it.

File: src/ray/gcs/redis_gcs_client.cc (hunk 9 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.cc]: It's unrelated to this PR. I can revert it.

File: src/ray/gcs/redis_gcs_client.h (hunk 18 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.h]: Not a TODO. Prompt everyone not to use the old interface.

File: src/ray/gcs/gcs_client.h [hunk 21]
```diff
@@ -7,12 +7,18 @@
...
 #include "ray/util/logging.h"
 
 namespace ray {
 
 namespace gcs {
 
+/// Specifies whether commands issued to a table should be regular or chain-replicated
+/// (when available).
+// TODO(micafan) Rename to RedisCommandType and maybe move it to gcs/type.h
+enum class CommandType { kRegular, kChain, kUnknown };
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client.h]: this is specific to redis gcs client and should be moved to that file?

File: src/ray/gcs/gcs_client.h [hunk 22]
```diff
@@ -87,6 +93,13 @@ class GcsClient : public std::enable_shared_from_this<GcsClient> {
     return *actor_accessor_;
   }
 
+  /// Get RedisJobInfoAccessor for reading or writing or subscribing to
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client.h]:
```suggestion
  /// Get JobInfoAccessor for reading or writing or subscribing to
```

File: src/ray/gcs/redis_gcs_client.h [hunk 23]
```diff
@@ -80,8 +81,10 @@ class RAY_EXPORT RedisGcsClient : public GcsClient {
...
+  JobTable &job_table();
```
Comment (zhijunfu) [inline src/ray/gcs/redis_gcs_client.h]: maybe just a RAY_FATAL with an error message

File: src/ray/gcs/redis_actor_info_accessor.cc [hunk 24]
```diff
@@ -92,7 +92,7 @@ Status RedisActorInfoAccessor::AsyncSubscribeAll(
     const SubscribeCallback<ActorID, ActorTableData> &subscribe,
     const StatusCallback &done) {
   RAY_CHECK(subscribe != nullptr);
-  return actor_sub_executor_.AsyncSubscribe(ClientID::Nil(), subscribe, done);
+  return actor_sub_executor_.AsyncSubscribeAll(ClientID::Nil(), subscribe, done);
```
Comment (zhijunfu) [inline src/ray/gcs/redis_actor_info_accessor.cc]: not specific to this PR, but is client_id required for AsyncSubscribeAll?

File: src/ray/raylet/node_manager.h [hunk 25]
```diff
@@ -390,12 +390,12 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   /// \return Void.
   void HandleObjectMissing(const ObjectID &object_id);
 
-  /// Handles updates to job table.
+  /// Handles job finished.
```
Comment (raulchen) [inline src/ray/raylet/node_manager.h]:
```suggestion
  /// Handles the event that a job is finished.
```

File: src/ray/gcs/test/redis_job_info_accessor_test.cc [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+namespace ray {
+
+namespace gcs {
+
+class RedisJobInfoAccessorTest : public AccessorTestBase<JobID, JobTableData> {
+ protected:
+  virtual void GenTestData() {
+    for (size_t i = 0; i < total_job_number_; ++i) {
+      JobID job_id = JobID::FromInt(i);
+      std::shared_ptr<JobTableData> job_data_ptr = std::make_shared<JobTableData>();
```
Comment (raulchen) [inline src/ray/gcs/test/redis_job_info_accessor_test.cc]: just use the helper function here?

File: src/ray/gcs/gcs_client.h (hunk 21 above)
Comment (micafan) [reply src/ray/gcs/gcs_client.h]: Sorry, I made a mistake.

File: src/ray/gcs/test/redis_job_info_accessor_test.cc (hunk 26 above)
Comment (micafan) [reply src/ray/gcs/test/redis_job_info_accessor_test.cc]: done

File: src/ray/raylet/node_manager.h (hunk 25 above)
Comment (micafan) [reply src/ray/raylet/node_manager.h]: done

File: src/ray/gcs/gcs_client.h [hunk 27]
```diff
@@ -87,6 +68,13 @@ class GcsClient : public std::enable_shared_from_this<GcsClient> {
     return *actor_accessor_;
   }
 
+  /// Get JobInfoAccessor for reading or writing or subscribing to
```
Comment (raulchen) [inline src/ray/gcs/gcs_client.h]:
```suggestion
  /// Get the sub-interface for accessing job information in GCS.
```
So it's aligned with `JobInfoAccessor`'s documentation.

File: src/ray/gcs/job_info_accessor.h [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+/// This class includes all the methods that are related to accessing
+/// job information in the GCS.
+class JobInfoAccessor {
+ public:
+  virtual ~JobInfoAccessor() = default;
+
+  /// Add a job to GCS asynchronously.
+  ///
+  /// \param data_ptr The job that will be add to GCS.
+  /// \param callback Callback that will be called after job has been registered
```
Comment (raulchen) [inline src/ray/gcs/job_info_accessor.h]:
```suggestion
  /// \param callback Callback that will be called after job has been added
```

File: src/ray/gcs/job_info_accessor.h [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+  ///
+  /// \param job_id ID of the job that will be make finished to GCS.
+  /// \param callback Callback that will be called after update finishes.
+  /// \return Status
+  virtual Status AsyncMarkFinished(const JobID &job_id,
+                                   const StatusCallback &callback) = 0;
+
+  /// Subscribe to finished jobs.
+  ///
+  /// \param subscribe Callback that will be called each time when a job finished.
```
Comment (raulchen) [inline src/ray/gcs/job_info_accessor.h]:
```suggestion
  /// \param subscribe Callback that will be called each time when a job finishes.
```

File: src/ray/gcs/redis_actor_info_accessor.cc (hunk 24 above)
Comment (raulchen) [reply src/ray/gcs/redis_actor_info_accessor.cc]: Then we can remove the client_id parameter from this function?

File: src/ray/gcs/redis_gcs_client.cc [hunk 30]
```diff
@@ -138,14 +148,13 @@ Status RedisGcsClient::Connect(boost::asio::io_service &io_service) {
...
+  job_accessor_.reset(new RedisJobInfoAccessor(this));
+
+  is_connected_ = true;
 
-  Status status = Attach(io_service);
-  is_connected_ = status.ok();
+  RAY_LOG(INFO) << "RedisGcsClient Connected.";
 
-  // TODO(micafan): Synchronously register node and look up existing nodes here
-  // for this client is Raylet.
```
Comment (raulchen) [inline src/ray/gcs/redis_gcs_client.cc]: This TODO doesn't look fixed?

**Review by raulchen — APPROVED** [review]
LGTM, only a few small comments.

File: src/ray/gcs/redis_gcs_client.cc (hunk 30 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.cc]: Plan is change, we will not register node in `GcsClient::Connect()`, but call the register method manually.

**Review by zhijunfu — APPROVED** [review]

File: src/ray/gcs/redis_gcs_client.cc (hunk 30 above)
Comment (raulchen) [reply src/ray/gcs/redis_gcs_client.cc]: ok

**Comment by raulchen** [comment]
Hmm, I don't know why CI is not running for this PR. Let me try pull master.

File: BUILD.bazel [hunk 31]
```diff
@@ -731,7 +732,18 @@ cc_binary(
...
-    srcs = ["src/ray/gcs/subscription_executor_test.cc"],
+    srcs = ["src/ray/gcs/test/subscription_executor_test.cc"],
+    copts = COPTS,
+    deps = [
+        ":gcs",
+        "@com_google_googletest//:gtest_main",
+    ],
+)
+
+cc_binary(
```
Comment (ericl) [inline BUILD.bazel]: Let's avoid adding any more cc_binary scripts that are used for testing. Instead, consider `cc_test` or `sh_test`.

File: src/ray/gcs/redis_gcs_client.cc [hunk 32]
```diff
@@ -73,7 +73,16 @@ namespace ray {
 
 namespace gcs {
 
-RedisGcsClient::RedisGcsClient(const GcsClientOptions &options) : GcsClient(options) {}
+RedisGcsClient::RedisGcsClient(const GcsClientOptions &options) : GcsClient(options) {
+#if RAY_USE_NEW_GCS
```
Comment (ericl) [inline src/ray/gcs/redis_gcs_client.cc]: You can safely remove NEW_GCS stuff, we won't need any of that.

File: src/ray/gcs/redis_gcs_client.cc [hunk 33]
```diff
@@ -114,6 +123,8 @@ Status RedisGcsClient::Connect(boost::asio::io_service &io_service) {
                                              /*password=*/options_.password_));
   }
 
+  Attach(io_service);
```
Comment (ericl) [inline src/ray/gcs/redis_gcs_client.cc]: Should we still check the status here?

File: src/ray/gcs/redis_job_info_accessor.h [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+namespace gcs {
+
+class RedisGcsClient;
+
+/// \class RedisJobInfoAccessor
+/// RedisJobInfoAccessor is an implementation of `JobInfoAccessor`
+/// that uses Redis as the backend storage.
+class RedisJobInfoAccessor : public JobInfoAccessor {
+ public:
+  explicit RedisJobInfoAccessor(RedisGcsClient *client_impl);
```
Comment (ericl) [inline src/ray/gcs/redis_job_info_accessor.h]: Avoid raw pointer here.

File: src/ray/gcs/redis_job_info_accessor.h [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+ private:
+  /// Append job information to GCS asynchronously.
+  ///
+  /// \param data_ptr The job information that will be appended to GCS.
+  /// \param callback Callback that will be called after append done.
+  /// \return Status
+  Status DoAsyncAppend(const std::shared_ptr<JobTableData> &data_ptr,
+                       const StatusCallback &callback);
+
+  RedisGcsClient *client_impl_{nullptr};
```
Comment (ericl) [inline src/ray/gcs/redis_job_info_accessor.h]: Avoid raw pointer.

**Review by ericl — CHANGES_REQUESTED** [review]
Main comment is that we should add proper unit tests instead of the scripted setup.

File: BUILD.bazel (hunk 31 above)
Comment (micafan) [reply BUILD.bazel]: Agreed. And I plan to do so in next PR [5515](https://github.com/ray-project/ray/pull/5515)

File: src/ray/gcs/redis_gcs_client.cc (hunk 33 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.cc]: Because the return status of `Attach` is always ok, so I change the return status to void.

File: src/ray/gcs/redis_job_info_accessor.h (hunk 34 above)
Comment (micafan) [reply src/ray/gcs/redis_job_info_accessor.h]: We can add a CHECK here if your concern is the pointer is not initialized.

File: src/ray/gcs/redis_gcs_client.cc (hunk 32 above)
Comment (micafan) [reply src/ray/gcs/redis_gcs_client.cc]: Sure. I will remove it in next PR 5515.

**Comment by raulchen** [comment]
@micafan Better use a clean PR to refactor the tests, instead mixing that with #5515. 

But I think it's fine to do that after this PR (and before merging any other similar PRs), as this PR is already ready. @ericl do you think this acceptable?

Also, CI failure is related
```
ERROR: missing input file '//:src/ray/gcs/test/redis_gcs_client_test.cc'
ERROR: /home/travis/build/ray-project/ray/BUILD.bazel:710:1: //:redis_gcs_client_test: missing input file '//:src/ray/gcs/test/redis_gcs_client_test.cc'
ERROR: /home/travis/build/ray-project/ray/BUILD.bazel:710:1 1 input file(s) do not exist
```

File: BUILD.bazel (hunk 31 above)
Comment (ericl) [reply BUILD.bazel]: Thanks, that sounds good.

File: src/ray/gcs/redis_job_info_accessor.h (hunk 34 above)
Comment (ericl) [reply src/ray/gcs/redis_job_info_accessor.h]: I see, in this case can we change `client_impl_` to a reference to indicate non-nullability?

**Review by ericl — APPROVED** [review]
Looks good, but consider using a reference instead of a raw pointer in the accessor class.
